lib/rust/flexer/src/automata/dfa.rs

```diff
@@ -10,52 +10,52 @@ use crate::data::matrix::Matrix;
+///
+/// ```text
+///  ┌───┐  'D'  ┌───┐  'F'  ┌───┐  'A'  ┌───┐
+///  │ 0 │──────▶│ 1 │──────▶│ 2 │──────▶│ 3 │
+///  └───┘       └───┘       └───┘       └───┘
+/// ```
+#[derive(Clone,Debug,Default,Eq,PartialEq)]
+pub struct DFA {
+    /// A set of disjoint intervals over the allowable input alphabet.
+    pub alphabet_segmentation: alphabet::Segmentation,
+    /// The transition matrix for the DFA.
+    ///
+    /// It represents a function of type `(state, symbol) -> state`, returning the identifier for
+    /// the new state.
+    ///
+    /// For example, the transition matrix for an automaton that accepts the language
+    /// `{"A" | "B"}*"` would appear as follows, with `-` denoting
+    /// [the invalid state](state::INVALID). The leftmost column encodes the input state, while the
+    /// topmost row encodes the input symbols.
+    ///
+    /// |   | A | B |
+    /// |:-:|:-:|:-:|
+    /// | 0 | 1 | - |
+    /// | 1 | - | 0 |
+    ///
+    pub links: Matrix<state::Identifier>,
+    /// A collection of callbacks for each state (indexable in order)
+    pub callbacks: Vec<Option<RuleExecutable>>,
 }
 
-/// DFA automata with a set of symbols, states and transitions.
-/// Deterministic Finite Automata is a finite-state machine that accepts or rejects a given sequence
-/// of symbols, by running through a state sequence uniquely determined by the input symbol sequence.
-///   ___              ___              ___              ___
-///  | 0 | -- 'D' --> | 1 | -- 'F' --> | 2 | -- 'A' --> | 3 |
-///   ‾‾‾              ‾‾‾              ‾‾‾              ‾‾‾
-/// More information at: https://en.wikipedia.org/wiki/Deterministic_finite_automaton
 
-#[derive(Clone,Debug,Default,PartialEq,Eq)]
-pub struct DFA {
-    /// Finite set of all valid input symbols.
-    pub alphabet: Alphabet,
-    /// Transition matrix of deterministic finite state automata.
-    /// It contains next state for each pair of state and input symbol - (state,symbol) => new state.
-    /// For example, a transition matrix for automata that accepts string "ABABAB...." would look
-    /// like this:
-    ///  states
-    /// |       | A | B | <- symbols
-    /// | 0     | 1 | - |
-    /// | 1     | - | 0 |
-    ///  Where `-` denotes `state::INVALID`.
-    pub links: Matrix<state::Id>,
-    /// Stores callback for each state (if it has one).
-    pub callbacks: Vec<Option<Callback>>,
-}
+// === Trait Impls ===
 
-impl From<Vec<Vec<usize>>> for Matrix<state::Id> {
+impl From<Vec<Vec<usize>>> for Matrix<state::Identifier> {
     fn from(input:Vec<Vec<usize>>) -> Self {
         let rows        = input.len();
         let columns     = if rows == 0 {0} else {input[0].len()};
         let mut matrix  = Self::new(rows,columns);
         for row in 0..rows {
             for column in 0..columns {
-                matrix[(row,column)] = state::Id{id:input[row][column]};
+                matrix[(row,column)] = state::Identifier {id:input[row][column]};
```

Comment:
I was writing about exactly this in other place of the code - use `Identifier::from(...)` instead. Basically having several `From` impls makes your code very versatile and easy to use. Definitely better than braces-based constructors (which we try to avoid) - in fact you will not find any braces-based constructors in ensogl code but in methods "new" :)